File: python-learning-path/practise/prime_check.py
"""
Exercise 03: Prime Number Checker

OPTIMIZATION:
Instead of checking all numbers up to n, we only check up to sqrt(n).
If n is divisible by any number, it's not prime.
"""

import math

def is_prime(n):
    # Prime numbers must be greater than 1
    if n <= 1:
        return False
    
    # Optimization: Loop from 2 up to the square root of n (inclusive)
    # Factors occur in pairs (e.g., for 36: 4*9, 6*6). One factor is always <= sqrt(n).
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False # Found a factor, not prime
            
    return True # No factors found, it's prime

def main():
    try:
        num = int(input("Enter a number to check: ") or "0")
        
        if is_prime(num):
            print(f"💎 {num} is a prime number.")
        else:
            print(f"⚪ {num} is NOT a prime number.")
            
    except ValueError:
        print("Invalid numeric input.")

if __name__ == "__main__":
    main()

File: java-learning-path/practise/PrimeCheck.java
import java.util.Scanner;

/**
 * Exercise 03: Prime Number Checker
 * 
 * ALGORITHM:
 * 1. A prime number is > 1.
 * 2. If it is divisible by any integer from 2 to the square root of itself, it's not prime.
 *    (Optimization: We only check up to sqrt(n) because factors come in pairs).
 */
public class PrimeCheck {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter a number: ");
        int num = scanner.nextInt();

        // 0 and 1 are not prime numbers
        boolean isPrime = num > 1;
        
        // Loop from 2 up to square root of num
        // Efficiency: If num has a factor larger than sqrt(num), it must also have one smaller than sqrt(num).
        for (int i = 2; i <= Math.sqrt(num); i++) {
            if (num % i == 0) {
                isPrime = false; // Found a factor, so not prime
                break;       // Exit loop early for efficiency
            }
        }

        if (isPrime) {
            System.out.println(num + " is a prime number.");
        } else {
            System.out.println(num + " is NOT a prime number.");
        }
        scanner.close();
    }
}
